File: go/138.copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    // mapping from orginal nodes to copied nodes
    m := make(map[*Node]*Node)
    // copy singly linked list
    newHead := deepCopy(head, m)
    // use the mapping to fill Random pointer
    for head != nil {
        m[head].Random = m[head.Random]
        head = head.Next
    }
    return newHead
}

// deep copy a singly linked list(ignore the Random pointer)
// and keep track of a mapping from original node to copied node
func deepCopy(head *Node, m map[*Node]*Node) *Node {
    if head == nil {
        return nil
    } else {
        next := deepCopy(head.Next, m)
        node := &Node{
            Val: head.Val,
            Next: next,
        }
        m[head] = node
        return node
    }
}

File: csharp/138.copy-list-with-random-pointer.cs
/*
 * @lc app=leetcode id=138 lang=csharp
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;
    
    public Node(int _val) {
        val = _val;
        next = null;
        random = null;
    }
}
*/

using System.Collections.Generic;

public class Solution 
{
    public Node CopyRandomList(Node head) 
    {
        Dictionary<Node, Node> mapping = new Dictionary<Node, Node>();

        Node cur, dummy, prev;
        dummy = new Node(int.MaxValue);
        prev = dummy;
        cur = head;

        // deep copy list
        while (cur != null)
        {
            prev.next = new Node(cur.val);
            mapping[cur] = prev.next;
            prev = prev.next;
            cur = cur.next;
        }

        // link random
        cur = head;
        while (cur != null)
        {
            if (cur.random == null)
            {
                mapping[cur].random = null;
            }
            else 
            {
                mapping[cur].random = mapping[cur.random];
            }
            cur = cur.next;
        }

        return dummy.next;
    }
}
// @lc code=end

